src/components/tl/self/tl_self_coll.c

```diff
@@ -0,0 +1,212 @@
+    return ucc_progress_queue_enqueue(UCC_TASK_CORE_CTX(task)->pq, task);
+}
+
+void ucc_tl_self_coll_copy_progress(ucc_coll_task_t *coll_task)
+{
+    ucc_tl_self_task_t *task = ucc_derived_of(coll_task, ucc_tl_self_task_t);
+
+    task->super.status = ucc_mc_memcpy(task->dst, task->src, task->size,
+                                       task->dst_memtype, task->src_memtype);
+}
+
+ucc_status_t ucc_tl_self_coll_start(ucc_coll_task_t *task)
+{
+    return ucc_progress_queue_enqueue(UCC_TASK_CORE_CTX(task)->pq, task);
+}
+
+ucc_status_t ucc_tl_self_coll_noop_init(ucc_tl_self_task_t *task)
+{
+    task->super.status   = UCC_OK;
+    task->super.post     = ucc_tl_self_coll_start;
+    task->super.progress = ucc_tl_self_noop_progress;
+    return UCC_OK;
+}
+
+ucc_status_t ucc_tl_self_coll_copy_init(ucc_tl_self_task_t *task)
+{
+    ucc_coll_args_t *args = &(task->super.bargs.args);
+
+    task->super.post = ucc_tl_self_coll_start;
+    if (UCC_IS_INPLACE(*args)) {
+        /* no copy is required for in-place */
+        task->super.progress = ucc_tl_self_noop_progress;
+    } else {
+        task->dst = args->dst.info.buffer;
+        task->src = args->src.info.buffer;
+        task->size =
+            args->src.info.count * ucc_dt_size(args->src.info.datatype);
+        task->dst_memtype    = args->dst.info.mem_type;
+        task->src_memtype    = args->src.info.mem_type;
+        task->super.progress = ucc_tl_self_coll_copy_progress;
+    }
+    return UCC_OK;
+}
+
+ucc_status_t ucc_tl_self_alltoallv_init(ucc_tl_self_task_t *task)
+{
+    ucc_coll_args_t *args = &(task->super.bargs.args);
+
+    task->super.post = ucc_tl_self_coll_start;
+    if (UCC_IS_INPLACE(*args)) {
+        /* no copy is required for in-place */
+        task->super.progress = ucc_tl_self_noop_progress;
+    } else {
+        size_t displ = (size_t)ucc_coll_args_get_displacement(
+            args, args->dst.info_v.displacements, 0);
+        task->dst = PTR_OFFSET(args->dst.info_v.buffer, displ);
+        displ     = (size_t)ucc_coll_args_get_displacement(
+            args, args->src.info_v.displacements, 0);
+        task->src  = PTR_OFFSET(args->src.info_v.buffer, displ);
+        task->size = ucc_coll_args_get_count(args, args->src.info_v.counts, 0) *
+                     ucc_dt_size(args->src.info_v.datatype);
+        task->dst_memtype    = args->dst.info_v.mem_type;
+        task->src_memtype    = args->src.info_v.mem_type;
+        task->super.progress = ucc_tl_self_coll_copy_progress;
+    }
+    return UCC_OK;
+}
+
+ucc_status_t ucc_tl_self_coll_copyv_init(ucc_tl_self_task_t *task)
+{
+    ucc_coll_args_t *args = &(task->super.bargs.args);
+
+    task->super.post = ucc_tl_self_coll_start;
+    if (UCC_IS_INPLACE(*args)) {
+        /* no copy is required for in-place */
+        task->super.progress = ucc_tl_self_noop_progress;
+    } else {
+        size_t displ = 0;
+        if (args->dst.info_v.displacements) {
```

Comment:
just wonder: this fn "copyv" is used for gatherv, allgatherv, and reduce_scatterv - doesn't this mean that correct app should always have "displacements"?